pubsublite/internal/test/verifier.go

```diff
@@ -0,0 +1,225 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+
+package test
+
+import (
+	"container/list"
+	"sync"
+	"testing"
+	"time"
+
+	"cloud.google.com/go/internal/testutil"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	// blockWaitTimeout is the timeout for any wait operations to ensure no
+	// deadlocks.
+	blockWaitTimeout = 30 * time.Second
+)
+
+type rpcMetadata struct {
+	wantRequest   interface{}
+	retResponse   interface{}
+	retErr        error
+	blockResponse chan struct{}
+}
+
+// wait until the `blockResponse` is released by the test, or a timeout occurs.
+// Returns immediately if there was no block.
+func (r *rpcMetadata) wait() error {
+	if r.blockResponse == nil {
+		return nil
+	}
+	select {
+	case <-time.After(blockWaitTimeout):
+		return status.Errorf(codes.FailedPrecondition, "mockserver: test did not unblock response within %v", blockWaitTimeout)
```

Comment:
I wonder if there's a better code here. Maybe `codes.Unavailable`?

Reply:
Yeah, DeadlineExceeded fits here too. The problem was that those are retryable codes, so the client would try to reconnect another stream, which would result in another error message, and it wasn't immediately obvious what was going wrong in the test. So I had to use FailedPrecondition, which was the most fitting of all the non-retryable codes.